feature(avatar): avatar is now served by the new API

hypeJunction opened the pull request:
Avatars are now served using elgg_get_inline_url()
- [x] Requires #9147

mrclay commented:
Looks good

hypeJunction commented:
This doesn't need a rebase. Just tested again manually, avatars are served as expected.

mrclay commented on mod/profile/icondirect.php:
```diff
@@ -6,6 +6,8 @@
  */
 
+elgg_deprecated_notice("icondirect.php is no longer used and will be removed, do not include and require it. Use elgg_get_inline_url() instead.", '2.1');
+
```
Should this just create the new image URL and forward to it?

hypeJunction replied:
Yes, that's what I did with groups. Will do that.

mrclay commented:
What if a plugin has changed the default user icons via `entity:icon:url` or the `resource/view/avatar` view? Won't this break that?

hypeJunction commented:
I don't see how. Their views/handlers will take precedence.

mrclay commented:
OK, yes I'm seeing that.

mrclay commented on engine/lib/users.php:
```diff
@@ -449,10 +449,16 @@ function user_avatar_hook($hook, $entity_type, $returnvalue, $params) {
 	$size = $params['size'];
 
 	if (isset($user->icontime)) {
-		return "avatar/view/$user->username/$size/$user->icontime";
```
If a plugin *only* overrides the view, their view will never be called.

mrclay replied:
Damn it, I was seeing this as a view render, not a URL. Sorry, ignore me.

mrclay replied:
No, that's right. If the default changes to not using the resource view, that's a problem for BC.

mrclay replied:
Ideally we could detect whether `resource/avatar/view` has been overridden and use it in that case.

hypeJunction replied:
Hmm, let me dig into it

hypeJunction replied:
We can leave the URL alone and forward from the resource view. Not sure how browsers will respond to caching with 301 headers...

mrclay replied:
In browsers where it didn't work it would be pretty bad. Seems risky.

mrclay replied:
This doesn't seem too tough: https://github.com/Elgg/Elgg/issues/9469#issuecomment-194861328

hypeJunction replied:
This particular view is not really extendable, it sends headers so if you output content before or after, you would be making a mess. We could just check view location. Trying that now.

hypeJunction replied:
Never mind, I see that we have to take into account the 'view' and 'view_vars' hooks...

hypeJunction commented:
icondirect now forwards to the new url

hypeJunction committed:
feature(avatar): user avatars are now served by serve-file handler

Avatars are now served using elgg_get_inline_url() via serve-file handler.
Profile plugin no longer overwrites user avatar URLs and avatar handlers.

hypeJunction commented:
@mrclay Updated to use `_elgg_view_may_be_altered()`

mrclay commented on engine/lib/users.php:
```diff
@@ -445,14 +445,28 @@ function user_create_hook_add_site_relationship($event, $object_type, $object) {
+	}
+
+	if (_elgg_view_may_be_altered('resources/avatar/view', 'resources/avatar/view.php')) {
 		return "avatar/view/$user->username/$size/$user->icontime";
```
We need a little comment "For BC with 2.0 if a plugin is suspected of using this view/page handler we need to use it."

hypeJunction replied:
Will add. FYI, tested manually by adding a resource view in a plugin. Didn't test with hooks.

hypeJunction replied:
Done

hypeJunction committed:
deprecate(groups): new file service deprecated avatar/view resource

avatar/view resource view and page handler have been deprecated along with
mod/profile/icondirect.php

hypeJunction commented:
Added a deprecation commit.

hypeJunction committed:
fix(avatars): avatars are no longer served with public URLs in a walled garden mode

Avatar URLs are now tied to a user session if walled garden mode is enabled.

Refs #9518

mrclay commented on docs/admin/upgrading.rst:
```diff
@@ -43,6 +43,12 @@ Deprecations in 2.x
 
+User avatars are now served via ``serve-file`` handler. Plugins should start using ``elgg_get_inline_url()`` and note that:
+
+ * `/avatar/view` page handler and resource view have been deprecated
```
Double backticks

mrclay replied:
Besides that LGTM

hypeJunction commented:
@mrclay updated docs to use double backticks
